Add `forbid_outofblock_ops()` to op.c; detect forbidden flow at compiletime of `defer` or `finally`

leonerd opened the pull request:
Add `forbid_outofblock_ops()` to op.c; detect forbidden flow at compiletime of `defer` or `finally`

Adds a new function to statically detect forbidden control flow out of a block.

Many kinds of forbidden control flow out of a `defer` or `finally` block can be detected statically with this function by analysing the optree, rather than leaving it until runtime when the attempt is actually made.

richardleach commented:
Sounds cool. Since the API function is intended to be generally useful, please could you put the function name in the commit message?

leonerd commented:
Commit message and PR description rewritten.

leonerd commented:
(I had to push an empty commit to force CI to retest stuff; for some reason it didn't rerun the tests and I don't know why. I'll remove it before merge)

leonerd committed:
Add `forbid_outofblock_ops()` to op.c

Adds a new function to statically detect forbidden control flow out of a
block.

leonerd committed:
Detect forbidden flow at compiletime of `defer` or `finally`

Using the new `forbid_outofblock_ops()`, many kinds of forbidden control
flow out of a `defer` or `finally` block can be detected statically with
this function by analysing the optree, rather than leaving it until
runtime when the attempt is actually made.

rwp0 reviewed (approved):
I don't claim to understand the implementation but the description (a switch from runtime to static analysis) is an improvement to the feature.

wolfsage commented:
Nice. Is it worth mentioning in the docs that "You should still have checks at run time" for cases like executing a subroutine or evaling a string that tries to goto out? That is, this static function alone isn't enough, it's just useful for finding compile time errors

leonerd commented:
> Is it worth mentioning in the docs that "You should still have checks at run time" for cases like executing a subroutine or evaling a string that tries to goto out?

Ahyes, definitely. I'll add some words to that effect.

rwp0 commented on pod/perldelta.pod:
```diff
@@ -66,6 +66,17 @@ C<INCDIR> methods will no longer trigger an exception, and instead will
+
+This version of perl adds compile-time detection for many cases that can be
+statically determined. This may mean that code which previously compiled
+successfully is now reported as an error, but only in cases where it would in
```
> but only in cases where it would in any case have been an error to actually execute the code anyway.

I re-read this sentence many times to understand it.
Perhaps, we can supply an example or leave just the main point?

leonerd replied:
I'll see if I can reword it somehow. Maybe:

"This version of perl adds compile-time detection for many cases that can be statically determined. This may mean that code which compiled successfully on a previous version of perl is now reported as a compile-time error with this one. This only happens in cases where it would in any case have been an error to actually execute the code anyway."

rwp0 replied:
> in cases where it would in any case have been an error to actually execute the code anyway.

The **clause** still remains unchanged :-) 

The new compile-time out-of-block error only happens where it was already an (runtime) error?

What's the meaning of this, why do we need to mention that, or maybe it's just me not getting the point yet (sorry, then) :-)

rwp0 replied:
I think we can say something like: out-of-block was already an error in previous Perl versions. But the error is now determined at compile-time rather than runtime as previously hence improving the speed of execution or error determination (if I got that right).

leonerd replied:
It's no faster to execute as we still have to detect those conditions anyway, because we can't *always* detect it statically. It just means that some kinds of programs that were always going to be an error are determined to be so at compiletime, rather than waiting until runtime to find out.

leonerd replied:
I've rewritten the words in perldelta now.

The user-facing reason is "because compiletime is better than runtime".

Additionally I wanted to use this new function in the feature-class branch, for protecting the expressions in `field $var = EXPR` so I needed the code around anyway. Since it's here (because I copied it from `Syntax::Keyword::Defer`) we might as well use it for `defer` and `finally` blocks too.
